### PR TITLE
Saving configurations at installation

### PR DESCRIPTION
**Pull request by thomasfangyi**
- Allow command line options to be saved into `configuration.json` and loaded when running `firedrake-update`
- Allow extra packages to be installed along firedrake with `--install PACKAGE` options. The packages will be tracked and updated when running `firedrake-update`
- <del>Fix a bug of cloning `slepc` again with an existing `slepc` directory when running `firedrake-update`</del> This has been moved to a separate PR

Ref #728

### CODE REVIEW

File: scripts/firedrake-install [hunk 1]
```diff
@@ -756,9 +813,7 @@ def build_update_script():
     with open("firedrake/scripts/firedrake-install", "r") as f:
         update_script = f.read()
 
-    for switch in ["developer", "sudo", "package_manager", "minimal_petsc",
-                   "mpicc", "disable_ssh", "show_petsc_configure_options", "adjoint",
-                   "virtualenv_name"]:
+    for switch in FiredrakeConfiguration._persistent_options:
         update_script = update_script.replace("args.%s = False" % switch,
                                               "args.%s = %r" % (switch, args.__getattribute__(switch)))
```
Comment (wence-) [inline scripts/firedrake-install]: This code just wants to be deleted we think.

File: scripts/firedrake-install [hunk 2]
```diff
@@ -140,13 +174,23 @@ else:
...
+
+    config = None
+    if not args.ignore_saved_config:
+        try:
+            import firedrake_configuration
+            config = firedrake_configuration.get_config()
+        except:
+            config = FiredrakeConfiguration(args)
+    if not config:
+        config = FiredrakeConfiguration(args)
```
Comment (wence-) [inline scripts/firedrake-install]: Command-line options should override saved configuration.  So after the config is built, one should parse the options into it. So we should do:

```python
class LegacyStage(object):
    pass
args = LegacyState()
args.developer = False
...

try:
   import firedrake_configuration
   config = firedrake_configuration.get_config()
except AppropriateError:
   # Get default config from legacy state
   config = FiredrakeConfiguration(args)

# Now we need to override stored config with commandline args
config.update(parser.parse_args())
```

File: scripts/firedrake-install (hunk 1 above)
Comment (thomasfangyi) [inline scripts/firedrake-install]: This is put back because it acts as a fall-back when reading/writing to an external JSON fails

**Comment by dham** [comment]
This could do with a test which checks that the configuration module loads and that the file can be read.

Currently if writing the configuration failed during install, this would not actually cause a build fail. Checking that it can be read afterwards overcomes this.

**Pull request merged**